conda_build/os_utils/macho.py

```diff
@@ -146,6 +146,23 @@ def _get_matching_load_commands(lines, cb_filter):
     return result
 
 
+def find_apple_cctools_executable(name, build_prefix, nofail=False):
+    tool = find_preferably_prefixed_executable(name, build_prefix)
+    try:
+        if '/usr/bin' in tool:
+            with open(tool, 'rb') as f:
+                s = f.read()
+            if s.find(b'usr/lib/libxcselect.dylib') != -1:
+                # This is not the real tool, but Apple's irritating GUI dialog launcher.
+                raise
+    except Exception as e:
+        print("ERROR :: Failed to run `{}`.  Please use `conda` to install `cctools` into your base environment.\n" \
```

Comment:
I should change this to say "into your build or base environment" I think?

Reply:
"please add cctools to your recipe's `build` section" is the nicest, IMHO

Reply:
but if that needs to be in effectively every recipe, then we'll be better off making it a run dep of conda-build on mac.

Reply:
or maybe not of conda-build, but of the compiler package on mac

Reply:
It is a dep of the compiler package! Packages that do not use compilers shouldn't be inflicted with them IMHO, not even in the base env.

Reply:
Is cctools necessary if people have no compiled code?  I appreciate separating them, I just wonder if we can exclude this.

Reply:
No, but there's no way to get into the code flow for making DSOs relative if you don't have macOS DSOs, and if you do have them (regardless of whether self-compiled or repackaged) then we need to be able to do the required path fixups. If you don't have DSOs then no need for compilers or binutilsy things, no, but you'd never see this message. This will only trip people who do need these things but haven't taken the necessary steps to provide them.

Reply:
Hmm, thinking about this, I think we should say something like "please use the `compiler` jinja2 function (link) as a build requirement or else, if you are sure you do not need compilers, add `cctools` instead".

.. or something worded better than that?

Reply:
I thought about it a bit more, and this absolutely has to be in the base env, NOT in the build env.  The build env is not available to conda-build for doing stuff with.  IMHO, we should just make this a CB dep on mac.

Reply:
I've not seen it not be available when needed so far, can you detail how that could happen? I'd rather not introduce our compiler decisions into our build tool so deeply if I can avoid it.

Reply:
This code is careful to *also* fall back to Apple's tools (which is in fact all we ever ran here before, with no warnings or information .. this is all improvement already, and if the build env is always available at this point, as I think it is, then I'd rather we talk about build envs here).

I actually wanted to go further and prefer our tools for everything conda-build needs (including things like git, but that's too difficult as you need git etc before you know your deps - for post stuff we already know them), but again, it would only *ever* be a preference change iff the tools are found where we look for them.

Reply:
Your post-build link checking stuff uses this, right?  At the stage where that happens, you can't assume that cctools is still around in the build env.  You shouldn't assume that the build env is around at all.  I have always just installed xcode right away, as most others will have done, too, I imagine.  I think this is for a better future world, where people won't need xcode, right?  In that world, if they only have cctools in their build env, then your tool will (should) fail to find cctools, unless you separately install it in the base env (or wherever conda-build lives or otherwise on PATH)

Reply:
So do note, anyone with Xcode or Xcode command line tools (currently everyone running conda-build) will not see this message. This frees us from needing Xcode.

Reply:
Yeah, we're saying the same thing, I think.  The message is fine.  Error out if people need to install that thing.  Their choices are either installing xcode or install our package.

Reply:
Also our activation scripts for cctools etc need to be free to set env vars should we decide to. Having env vars in the base env for our core tools is a line I'd like not to cross.

Reply:
(full disclosure, our clang package sets these env vars for now, but that's because I was lazy!)

Reply:
.. and yes, we could indirect it through a cctools activation (or implementation) package to avoid this problem.

Reply:
One quick thing, should we have different exceptions for these errors? BuildToolMissing, BuildToolIsAppleStub?

Reply:
Under what condition can we not assume it's around. Post to me implies it gets run just after we do the build, and if any env destruction happened in-between then that would seem incorrect to me. Now moving build before test is totally fine and to be expected.